internal/api/external.go

```diff
@@ -65,6 +69,23 @@ func (a *API) ExternalProviderRedirect(w http.ResponseWriter, r *http.Request) e
 	log := observability.GetLogEntry(r)
 	log.WithField("provider", providerType).Info("Redirecting to external provider")
 
+	oauthID := ""
+	if flowType == PKCE {
+		codeChallenge := query.Get("code_challenge")
+		if codeChallenge == "" {
+			return badRequestError("Code challenge must be non-empty in pkce flow")
+		}
+		oauthState, err := models.NewOAuthState(providerType, codeChallenge)
+		if err != nil {
+			return err
+		}
+		if err := a.db.Create(oauthState); err != nil {
+			return err
+		}
+
+		oauthID = oauthState.ID.String()
```

Comment:
There are a couple of more things to validate here:

1. Check that the code_challenge is not very short, otherwise it makes no sense. Check the PKCE spec if it asks for a minimum length or format?
2. To be fully compliant with the spec, there was a property that tells us what the transformation is. If it is not included, it's SHA-256, but if it is make sure only SHA-256 is accepted.